File: src/codility/java/leader/Dominator.java
package codility.java.leader;

import java.util.HashMap;
import java.util.Map;

public class Dominator {

    public int solution(int[] A) {
        Map<Integer, Integer> map = new HashMap<>();

        double max = 0;

        for (int i =0;i<A.length;i++) {
            int a = A[i];
            int count = map.getOrDefault(a, 0) + 1;
            map.put(a, count);
            max = Math.max(count, max);
            if (max > A.length / 2D) {
                return i;
            }
        }

        return -1;

    }
}

File: src/codility/kotlin/Dominator.kt
package codility.kotlin

class Dominator {

    fun solution(A: IntArray): Int {
        var leader = A[0]
        var leaderCount = 1
        var leaderIndex = 0

        for (i in 1 until A.size) {
            if (leaderCount == 0) {
                leaderCount = 1
                leader = A[i]
                leaderIndex = i
            } else if (A[i] == leader) {
                leaderCount++
            } else {
                leaderCount--
            }
        }

        var count = 0
        for (a in A) {
            if (a == leader) {
                count++
            }
        }
        return if (count > A.size.toDouble() / 2) leaderIndex else -1

    }
}